### PR TITLE
Update OAuth tokens every time a user signs in

### PR DESCRIPTION
**Pull request by gabebw**
Croniker is having a problem where it periodically can't authenticate users and throws errors (despite having OAuth tokens set), but manually setting OAuth tokens in the DB fixes it.

I suspect this is because OAuth tokens are somehow expiring, maybe because they're reset when a user signs in. Since Croniker expires sessions after one week, this theory means that OAuth tokens are expiring about every week for some frequent users (me).

Previously Croniker would set tokens when a user signed up, but then ignore new tokens every time after that.

To fix this, Croniker now resets every user's OAuth tokens to be the current tokens every time they sign in.

### CODE REVIEW

File: Model/User.hs [hunk 1]
```diff
@@ -14,12 +14,17 @@ authenticateUser Creds{credsExtra} = do
...
-        (Just user) -> return $ Authenticated $ entityKey user
+        (Just (Entity userId _)) -> updateUser userId
     where
         createUser :: AuthId m ~ UserId => Maybe User -> DB (AuthenticationResult m)
         createUser (Just user) = Authenticated <$> insert user
         createUser Nothing = return $ ServerError "Something went wrong"
 
+        updateUser :: (AuthId m ~ UserId) => UserId -> DB (AuthenticationResult m)
+        updateUser userId = do
+            update userId (updateStatements credsExtra)
```
Comment (gabebw) [inline Model/User.hs]: Unfortunately this returns `m ()` so there's no way to check if it succeeded.

Comment (tonyd256) [inline Model/User.hs]: It's haskell ... it can't fail

Comment (gabebw) [inline Model/User.hs]: The database isn't written in Haskell, though!

File: Model/User.hs [hunk 2]
```diff
@@ -29,5 +34,11 @@ credsToUser credsExtra = User
     <*> pure defaultTZLabel
     <*> pure False
 
+updateStatements :: [(Text, Text)] -> [Update User]
+updateStatements credsExtra = catMaybes [
+        (UserTwitterOauthToken =.) <$> (lookup "oauth_token" credsExtra),
+        (UserTwitterOauthTokenSecret =.) <$> (lookup "oauth_token_secret" credsExtra)
+    ]
```
Comment (tonyd256) [inline Model/User.hs]: typically Haskellers like the format like:
```
    catMaybes
        [ firstThing
        , secondThing
        ]
```
not sure if you're following that .. if not, ignore me

File: Model/User.hs [hunk 3]
```diff
@@ -14,12 +14,17 @@ authenticateUser Creds{credsExtra} = do
     muser <- maybe (return Nothing) (getBy . UniqueUser) mTwitterUserId
     case muser of
         Nothing -> createUser (credsToUser credsExtra)
-        (Just user) -> return $ Authenticated $ entityKey user
+        (Just (Entity userId _)) -> updateUser userId
```
Comment (tonyd256) [inline Model/User.hs]: don't think you need the outer parens here

File: Model/User.hs [hunk 4]
```diff
@@ -29,5 +34,11 @@ credsToUser credsExtra = User
     <*> pure defaultTZLabel
     <*> pure False
 
+updateStatements :: [(Text, Text)] -> [Update User]
+updateStatements credsExtra = catMaybes [
+        (UserTwitterOauthToken =.) <$> (lookup "oauth_token" credsExtra),
```
Comment (tonyd256) [inline Model/User.hs]: I don't think you need the parens around the stuff after `<$>`

File: Model/User.hs (hunk 1 above)
Comment (tonyd256) [inline Model/User.hs]: I'm sure `DB` has `ExceptT` or `IO` in it to completely crash your app if anything goes wrong so no worries

Comment (gabebw) [inline Model/User.hs]: > it can't fail

:+1:

File: Model/User.hs (hunk 2 above)
Comment (gabebw) [inline Model/User.hs]: I like following convention!

Comment (gabebw) [inline Model/User.hs]: So, yes, I'll do that.

**Comment by gabebw** [comment]
Thanks for the review, @tonyd256!

**Pull request merged**